framework/doc/content/source/variables/ArrayMooseVariable.md

```diff
@@ -0,0 +1,89 @@
+OutputShape is for the type of shape functions and OutputData is the type of basis function expansion coefficients that are stored in the solution vector.
+
+The final form of an array kernel is quite simple. Using `ArrayDiffusion` as an example. The `computeQpResidual` function has
+
+```
+  return _grad_u[_qp] * _array_grad_test[_i][_qp] * (*_d)[_qp];
+```
+
+where `_grad_u[_qp]` is an `Eigen::Matrix` with number of rows being equal to the number of components of the array variable and number of columns being `LIBMESH_DIM`. `_array_grad_test[_i][_qp]` is an `Eigen::Map` of classic `_grad_test[_i][_qp]` which is in type of `Gradient`. Thanks to the Eigen matrix arithmetic operators, we can have a simple multiplication expression here. `_d` is a pointer of a material property of `Real` type for scalar diffusion coefficient. Here we assume the diffusion coefficient is the same for all components. The returned value is RealArrayValue, i.e. an Eigen vector.
+
+Of course, if we have different diffusion coefficients for different components, we will have something like
+
+```
+    RealArrayValue v = _grad_u[_qp] * _array_grad_test[_i][_qp];
+    for (unsigned int i = 0; i < _var.count(); ++i)
+      v(i) *= (*_d_array)[_qp](i);
+```
+
+`_var.count()` gives the number of components of the array variable.
+If we have coupled diffusion terms, i.e. diffusion coefficient is a matrix, we will have
+
+```
+  return (*_d_2d_array)[_qp] * (_grad_u[_qp] * _array_grad_test[_i][_qp]);
+```
+
+Correspondingly the `computeQpJacobian` has
+
+```
+    return RealArrayValue::Constant(_var.count(),
+                                    _grad_phi[_j][_qp] * _grad_test[_i][_qp] * (*_d)[_qp]);
+```
+
+for a scalar diffusion coefficient or
+
+```
+    return _grad_phi[_j][_qp] * _grad_test[_i][_qp] * (*_d_array)[_qp];
+```
+
+for an array diffusion coefficient or
+
+```
+    return _grad_phi[_j][_qp] * _grad_test[_i][_qp] * (*_d_2d_array)[_qp].diagonal();
+```
+
+for a matrix diffusion coefficient.
+
+It is noted that only the diagonal entries of the diffusion coefficients are used in the fully-coupled case because `computeQpJacobian` is supposed to only assemble the block-diagonal part of the Jacobian.
+The full local Jacobian is assembled in function `computeQpOffDiagJacobian`, where when the off-diagonal variable is the array variable, we have
+
+```
+  return _grad_phi[_j][_qp] * _grad_test[_i][_qp] * (*_d_2d_array)[_qp];
+```
+
+The retuned value is in type of an Eigen matrix with number of rows and columns equal to the number of components.
+
+Future work:
+
+- To profile the code with large number of variables to find the hot spots and fix them if there are any.
+- To change the current dof ordering for elemental variables so that we can avoid bunch of if statements with `isNodal()` in `MooseVariableFE.C`, (refer to [libMesh Issue 2114](https://github.com/libMesh/libmesh/issues/2114).
```

Comment:
That would be fantastic. Can you open a ticket for this after this goes in?

Reply:
Will do.